// dbuf/dist/DReadBuf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DReadBuf = /** @class */ (function () {
    function DReadBuf(buf) {
        this.buf = buf;
        this.offset = 0;
    }
    DReadBuf.prototype.readBool = function () {
        var v = this.buf.readInt8(this.offset) ? true : false;
        this.offset += 1;
        return v;
    };
    DReadBuf.prototype.readInt32 = function () {
        var v = this.buf.readInt32LE(this.offset);
        this.offset += 4;
        return v;
    };
    DReadBuf.prototype.readFloat32 = function () {
        var v = this.buf.readFloatLE(this.offset);
        this.offset += 4;
        return v;
    };
    DReadBuf.prototype.readJSON = function () {
        var v = this.readString();
        if (!v) {
            return;
        }
        try {
            v = JSON.parse(v);
        }
        catch (e) {
            console.log(e);
        }
        return v;
    };
    DReadBuf.prototype.readString = function () {
        var len = this.readInt32();
        var tmp = this.buf.toString("utf8", this.offset, this.offset + len);
        this.offset += len;
        return tmp;
    };
    DReadBuf.prototype.readBuffer = function () {
        var len = this.readInt32();
        var tmp = Buffer.allocUnsafe(len);
        Buffer.from(this.buf, this.offset, len);
        this.buf.copy(tmp, 0, this.offset, this.offset + len);
        this.offset += len;
        return tmp;
    };
    return DReadBuf;
}());
exports.default = DReadBuf;
